tests/python_package_test/test_engine.py

```diff
@@ -3303,6 +3303,131 @@ def test_force_split_with_feature_fraction(tmp_path):
         assert tree_structure['split_feature'] == 0
 
 
+def test_goss_boosting_and_strategy_equivalent():
+    X, y = make_regression(n_samples=10_000, n_features=10, n_informative=5, random_state=42)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    lgb_train = lgb.Dataset(X_train, y_train)
+    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
+    params1 = {
+        'boosting': 'goss',
+        'metric': 'l2',
+        'verbose': -1
+    }
+    evals_result1 = {}
+    gbm = lgb.train(params1, lgb_train,
+                    num_boost_round=10,
+                    valid_sets=lgb_eval,
+                    callbacks=[lgb.record_evaluation(evals_result1)])
+    params2 = {
+        'data_sample_strategy': 'goss',
+        'metric': 'l2',
+        'verbose': -1
+    }
+    evals_result2 = {}
+    gbm = lgb.train(params2, lgb_train,
+                    num_boost_round=10,
+                    valid_sets=lgb_eval,
+                    callbacks=[lgb.record_evaluation(evals_result2)])
+    np.testing.assert_allclose(evals_result1['valid_0']['l2'], evals_result2['valid_0']['l2'])
+
+
+def test_sample_strategy_with_boosting():
```

Comment:
I believe we should assert that `ret`s are **identical** for cases in which they should be **and not identical** otherwise.

Reply:
I agree with that, I think it should be possible to eliminate all sources of randomness here so that an exact result can be tested for. That would be useful for catching performance regressions in the future (https://github.com/microsoft/LightGBM/pull/4228#issuecomment-830706993, https://github.com/microsoft/LightGBM/pull/4228#issuecomment-862710227)

Reply:
@jameslamb Thanks for your comment, but it seems we are speaking about a little bit different things.

My proposal was about something like that:
```python
assert ret1 == ret2
assert ret1 != ret3
assert ret1 == ret5
...
```

If I understood correctly, you propose something like the following:
```
assert ret1 == pytest.approx(3954)
assert ret3 == pytest.approx(3800)
...
```

I believe both ideas are correct and important, but for "an exact result can be tested for" it will be better to assign expected result to a variable first and then use it in asserts with the aim to eliminate places in the code that potentially might be modified in the future in a result of changing something in the core LightGBM algorithms.

Reply:
oh, I see! Yes I agree with you.

Reply:
It is weird that both training with `params1` and `params2` in the test case `test_goss_boosting_and_strategy_equivalent` result in slightly different values in each run.

```
run 0
params1, params2
{9872.34900920269137714058, 9872.34900920268773916177}
{8357.18941658444782660808, 8357.18941658444964559749}
{7120.98798096409245772520, 7120.98798096409245772520}
{6077.75160468498961563455, 6077.75160468499052512925}
{5184.25930526859337987844, 5184.25930526859156088904}
{4470.25222677429155737627, 4470.25222677429064788157}
{3865.26989708719020200078, 3865.26989708719065674813}
{3360.59952312780615102383, 3360.59952312780569627648}
{2913.05571221191985387122, 2913.05571221192030861857}
{2547.71595821537584924954, 2547.71595821537630399689}

run 1
params1, params2
{9872.34900920269319612999, 9872.34900920269137714058}
{8357.18941658444782660808, 8357.18941658444964559749}
{7120.98798096409336721990, 7120.98798096409518620931}
{6077.75160468498961563455, 6077.75160468498961563455}
{5184.25930526859247038374, 5184.25930526859247038374}
{4470.25222677429155737627, 4470.25222677429155737627}
{3865.26989708719111149549, 3865.26989708719111149549}
{3360.59952312780751526589, 3360.59952312780569627648}
{2913.05571221192076336592, 2913.05571221191985387122}
{2547.71595821537539450219, 2547.71595821537584924954}
```
I even set `bagging_seed=0, seed=0, deterministic=True, force_col_wise=True`. But there's no luck.

Reply:
Maybe worth checking with `num_threads=1`?
Refer to #2804.

Reply:
Or maybe there are some float/double mismatches?..

Reply:
> Maybe worth checking with num_threads=1

I've tried that but no luck.
I think this is an issue beyond the content of this PR. Maybe we can use `assert_allclose` temporarily and investigate the reason laster on.

Reply:
> Maybe we can use `assert_allclose` temporarily and investigate the reason laster on.

OK, I see. Works for me. Just please don't forget to create a new issue for this task to not lose it deep inside PR's conversations.

Reply:
@shiyu1994 
> I think this is an issue beyond the content of this PR. 

Hmmm, are you sure in this statement?

I just run the following code with d4cdbcfe1b1db2d48e95665a8b358028c496cdb2 and it seems that results don't varying from run to run in the `master` branch.

```python
import numpy as np
import lightgbm as lgb

from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

ret = []
for _ in range(1000):
    X, y = make_regression(n_samples=10_000, n_features=10, n_informative=5, random_state=42)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
    lgb_train = lgb.Dataset(X_train, y_train)
    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
    params1 = {
        'boosting': 'goss',
        'metric': 'l2',
        'verbose': -1
    }
    evals_result1 = {}
    gbm = lgb.train(params1, lgb_train,
                    num_boost_round=10,
                    valid_sets=lgb_eval,
                    callbacks=[lgb.record_evaluation(evals_result1)])
    params2 = {
        'boosting': 'goss',
        'metric': 'l2',
        'verbose': -1
    }
    evals_result2 = {}
    gbm = lgb.train(params2, lgb_train,
                    num_boost_round=10,
                    valid_sets=lgb_eval,
                    callbacks=[lgb.record_evaluation(evals_result2)])
    ret.append(evals_result1['valid_0']['l2'])
    ret.append(evals_result2['valid_0']['l2'])

ret = np.array(ret)
assert (ret == ret[0]).all()
```

Given that `'boosting': 'goss'` and `'data_sample_strategy': 'goss'` are now just two aliases of the same thing, I suspect there is something wrong with reproducibility in this PR.

Reply:
Thanks for the verification! Let me check in depth.

Reply:
Did you a default number of thread 1? I've tried with commit https://github.com/microsoft/LightGBM/commit/d4cdbcfe1b1db2d48e95665a8b358028c496cdb2. With `num_threads=16`, the result is still inconsistent. With `num_threads=1` we can get exactly the same outputs every time.
BTW, I found that `num_threads=1` does work for this PR (I forgot to change both `num_threads` to 1 in both params. Sorry about that).
In addition, I've adjust the learning rate from to `0.05`, with a default learning rate (0.1) and `num_boost_round=10`, the GOSS is not triggered actually, see
https://github.com/microsoft/LightGBM/blob/d4cdbcfe1b1db2d48e95665a8b358028c496cdb2/src/boosting/goss.hpp#L154-L157

Reply:
> Did you a default number of thread 1?

Nope, I used default `-1` value.

Just run the [same code](https://github.com/microsoft/LightGBM/pull/4827#discussion_r830538558) but with `'num_threads': 4` added in `params1/2` (my CPU is 4-core) and the assert for equality of results from all runs is passed again. Quite strange.

@jameslamb Could you please help to runt the referenced snippet on your local machine? Also, we can run it on CI to cover all OSes.

Reply:
I can run this on my Mac in the next day or two.

Reply:
I created a modified version of the sample code from https://github.com/microsoft/LightGBM/pull/4827#discussion_r830538558 which would allow setting number of threads for `params` in an environment variable (i.e. not relying on `OMP_NUM_THREADS`), and put that in a script.

<details><summary>test-pr-4827.py (click me)</summary>

```python
import os
import json
import numpy as np
import lightgbm as lgb

from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

ret = []
for i in range(1, 1001):
    if i % 50 == 0:
        print(f"attempt {i}")
    X, y = make_regression(n_samples=10_000, n_features=10, n_informative=5, random_state=42)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
    lgb_train = lgb.Dataset(X_train, y_train)
    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)
    params1 = {
        'boosting': 'goss',
        'metric': 'l2',
        'verbose': -1,
        'num_thread': int(os.getenv("LGB_THREADS", "-1"))
    }
    evals_result1 = {}
    gbm = lgb.train(params1, lgb_train,
                    num_boost_round=10,
                    valid_sets=lgb_eval,
                    callbacks=[lgb.record_evaluation(evals_result1)])
    params2 = {
        'boosting': 'goss',
        'metric': 'l2',
        'verbose': -1,
        'num_thread': int(os.getenv("LGB_THREADS", "-1"))
    }
    evals_result2 = {}
    gbm = lgb.train(params2, lgb_train,
                    num_boost_round=10,
                    valid_sets=lgb_eval,
                    callbacks=[lgb.record_evaluation(evals_result2)])
    ret.append(evals_result1['valid_0']['l2'])
    ret.append(evals_result2['valid_0']['l2'])

with open("out.json", "w") as f:
    f.write(json.dumps(ret, indent=4))

ret = np.array(ret)

assert (ret == ret[0]).all()
```

</details>

Then I tried running it for different values of `num_thread`, on latest `master` and the latest commit of this PR.

| commit                                              | `num_thread` | results?  |
|:---------------------------------------------------:|:------------:|:---------:|
| `master` (17d4e007852b4f0b34b211a674cb26e6751c98e2) | 1            | identical |
| `master` (17d4e007852b4f0b34b211a674cb26e6751c98e2) | 2            | identical |
| `master` (17d4e007852b4f0b34b211a674cb26e6751c98e2) | 4            | different |
| `master` (17d4e007852b4f0b34b211a674cb26e6751c98e2) | -1           | different |
| this PR (1e4c11a8f3f5d5f969627387a76c44d924f28464)  | 1            | identical |
| this PR (1e4c11a8f3f5d5f969627387a76c44d924f28464)  | 2            | identical |
| this PR (1e4c11a8f3f5d5f969627387a76c44d924f28464)  | 4            | different |
| this PR (1e4c11a8f3f5d5f969627387a76c44d924f28464)  | -1           | different |

I got the same results on `master` and this PR...slight precision differences in `evals_result` when `num_threads` was greater than the number of physical CPUs.

<hr>

### Details

After installing `lightgbm` from source, I ran that script like this:

```shell
LGB_THREADS=1 \
python "${HOME}/test-pr-4827.py"

LGB_THREADS=2 \
python "${HOME}/test-pr-4827.py"

LGB_THREADS=4 \
python "${HOME}/test-pr-4827.py"

# use default number of threads
LGB_THREADS=-1 \
python "${HOME}/test-pr-4827.py"
```

I also tested the existing OpenMP details in my environment as follows, using the advice from @ogrisel in https://github.com/microsoft/LightGBM/issues/4632#issuecomment-937834608 (thanks again for that, Olivier!!).

```shell
python \
    -m threadpoolctl \
    -i lightgbm
```

```text
[
  {
    "user_api": "blas",
    "internal_api": "openblas",
    "prefix": "libopenblas",
    "filepath": "/Users/jlamb/mambaforge/lib/libopenblasp-r0.3.18.dylib",
    "version": "0.3.18",
    "threading_layer": "openmp",
    "architecture": "Haswell",
    "num_threads": 4
  },
  {
    "user_api": "openmp",
    "internal_api": "openmp",
    "prefix": "libomp",
    "filepath": "/Users/jlamb/mambaforge/lib/libomp.dylib",
    "version": null,
    "num_threads": 4
  },
  {
    "user_api": "openmp",
    "internal_api": "openmp",
    "prefix": "libomp",
    "filepath": "/usr/local/Cellar/libomp/13.0.1/lib/libomp.dylib",
    "version": null,
    "num_threads": 4
  }
]
```

<details><summary>how I installed from master (click me)</summary>

```shell
git clone \
    --recursive \
    git@github.com:microsoft/LightGBM.git \
    lgb-main

cd ./lgb-main
git checkout 17d4e007852b4f0b34b211a674cb26e6751c98e2

cd ./python-package
python setup.py install
```

</details>

<details><summary>how I installed from this PR (click me)</summary>

```shell
git clone \
    --recursive \
    git@github.com:GuangdaLiu/LightGBM.git \
    lgb-decouple

cd ./lgb-decouple
git fetch origin decouple
git checkout decouple
git checkout 1e4c11a8f3f5d5f969627387a76c44d924f28464
git submodule update --recursive

cd ./python-package
python setup.py install
```

</details>

I'm running macOS Monterey (12.2.1), with two dual-core Intel i5 CPUs.

<details><summary>more environment details (click me)</summary>

Output of `conda info`.

```text
     active environment : None
       user config file : /Users/jlamb/.condarc
 populated config files : /Users/jlamb/mambaforge/.condarc
                          /Users/jlamb/.condarc
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.7.final.0
       virtual packages : __osx=12.2.1=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /Users/jlamb/mambaforge  (writable)
      conda av data dir : /Users/jlamb/mambaforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/jlamb/mambaforge/pkgs
                          /Users/jlamb/.conda/pkgs
       envs directories : /Users/jlamb/mambaforge/envs
                          /Users/jlamb/.conda/envs
               platform : osx-64
             user-agent : conda/4.11.0 requests/2.27.1 CPython/3.9.7 Darwin/21.3.0 OSX/12.2.1
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

Output of `conda env export --name base`.

```text
name: base
channels:
  - conda-forge
  - defaults
dependencies:
  - appnope=0.1.2=py39h6e9494a_2
  - argon2-cffi=21.3.0=pyhd8ed1ab_0
  - argon2-cffi-bindings=21.2.0=py39h89e85a6_1
  - asttokens=2.0.5=pyhd8ed1ab_0
  - attrs=21.4.0=pyhd8ed1ab_0
  - backcall=0.2.0=pyh9f0ad1d_0
  - backports=1.0=py_2
  - backports.functools_lru_cache=1.6.4=pyhd8ed1ab_0
  - beautifulsoup4=4.10.0=pyha770c72_0
  - black=22.1.0=pyhd8ed1ab_0
  - bleach=4.1.0=pyhd8ed1ab_0
  - brotlipy=0.7.0=py39h89e85a6_1003
  - bzip2=1.0.8=h0d85af4_4
  - c-ares=1.18.1=h0d85af4_0
  - ca-certificates=2021.10.8=h033912b_0
  - certifi=2021.10.8=py39h6e9494a_1
  - cffi=1.15.0=py39he338e87_0
  - charset-normalizer=2.0.10=pyhd8ed1ab_0
  - click=8.0.4=py39h6e9494a_0
  - conda=4.11.0=py39h6e9494a_0
  - conda-package-handling=1.7.3=py39h89e85a6_1
  - cpplint=1.6.0=pyhd8ed1ab_0
  - cryptography=36.0.1=py39h209aa08_0
  - dataclasses=0.8=pyhc8e2a94_3
  - dbus=1.13.6=h811a1a6_3
  - debugpy=1.5.1=py39h9fcab8e_0
  - decorator=5.1.1=pyhd8ed1ab_0
  - defusedxml=0.7.1=pyhd8ed1ab_0
  - entrypoints=0.4=pyhd8ed1ab_0
  - executing=0.8.3=pyhd8ed1ab_0
  - expat=2.4.7=h96cf925_0
  - flake8=4.0.1=pyhd8ed1ab_1
  - flit-core=3.7.1=pyhd8ed1ab_0
  - gettext=0.19.8.1=hd1a6beb_1008
  - icu=69.1=he49afe7_0
  - idna=3.3=pyhd8ed1ab_0
  - importlib-metadata=4.2.0=py39h6e9494a_0
  - importlib_resources=5.4.0=pyhd8ed1ab_0
  - iniconfig=1.1.1=pyh9f0ad1d_0
  - ipykernel=6.9.1=py39h71a6800_0
  - ipython=8.1.1=py39h6e9494a_0
  - ipython_genutils=0.2.0=py_1
  - ipywidgets=7.6.5=pyhd8ed1ab_0
  - isort=5.10.1=pyhd8ed1ab_0
  - jedi=0.18.1=py39h6e9494a_0
  - jinja2=3.0.3=pyhd8ed1ab_0
  - joblib=1.1.0=pyhd8ed1ab_0
  - jpeg=9e=h0d85af4_0
  - jsonschema=4.4.0=pyhd8ed1ab_0
  - jupyter=1.0.0=py39h6e9494a_7
  - jupyter_client=7.1.2=pyhd8ed1ab_0
  - jupyter_console=6.4.3=pyhd8ed1ab_0
  - jupyter_core=4.9.2=py39h6e9494a_0
  - jupyterlab_pygments=0.1.2=pyh9f0ad1d_0
  - jupyterlab_widgets=1.0.2=pyhd8ed1ab_0
  - krb5=1.19.2=hcfbf3a7_3
  - libarchive=3.5.2=h2b60450_1
  - libblas=3.9.0=13_osx64_openblas
  - libcblas=3.9.0=13_osx64_openblas
  - libclang=13.0.1=default_he082bbe_0
  - libcurl=7.81.0=hf45b732_0
  - libcxx=12.0.1=habf9029_1
  - libedit=3.1.20191231=h0678c8f_2
  - libev=4.33=haf1e3a3_1
  - libffi=3.4.2=h0d85af4_5
  - libgfortran=5.0.0=9_3_0_h6c81a4c_23
  - libgfortran5=9.3.0=h6c81a4c_23
  - libglib=2.70.2=hf1fb8c0_4
  - libiconv=1.16=haf1e3a3_0
  - liblapack=3.9.0=13_osx64_openblas
  - libllvm13=13.0.1=h64f94b2_2
  - libmamba=0.19.1=h9ce2d19_0
  - libmambapy=0.19.1=py39h2c464dc_0
  - libnghttp2=1.46.0=h6f36284_0
  - libopenblas=0.3.18=openmp_h3351f45_0
  - libpng=1.6.37=h7cec526_2
  - libpq=14.2=hea3049e_0
  - libsodium=1.0.18=hbcb3906_1
  - libsolv=0.7.19=hcf210ce_5
  - libssh2=1.10.0=h52ee1ee_2
  - libxml2=2.9.12=h7e28ab6_1
  - libzlib=1.2.11=h9173be1_1013
  - llvm-openmp=12.0.1=hda6cdc1_1
  - lz4-c=1.9.3=he49afe7_1
  - lzo=2.10=haf1e3a3_1000
  - mamba=0.19.1=py39ha435c47_0
  - markupsafe=2.1.0=py39h63b48b0_0
  - matplotlib-inline=0.1.3=pyhd8ed1ab_0
  - mccabe=0.6.1=py_1
  - mistune=0.8.4=py39h89e85a6_1005
  - mypy=0.931=py39h89e85a6_2
  - mypy_extensions=0.4.3=py39h6e9494a_4
  - mysql-common=8.0.28=h694c41f_0
  - mysql-libs=8.0.28=h115446f_0
  - nbclient=0.5.13=pyhd8ed1ab_0
  - nbconvert=6.4.4=py39h6e9494a_0
  - nbformat=5.2.0=pyhd8ed1ab_0
  - nbqa=1.3.1=pyhd8ed1ab_0
  - ncurses=6.2=h2e338ed_4
  - nest-asyncio=1.5.4=pyhd8ed1ab_0
  - notebook=6.4.8=pyha770c72_0
  - nspr=4.32=hcd9eead_1
  - nss=3.74=h31e2bf1_0
  - numpy=1.22.0=py39h9d9ce41_0
  - openssl=1.1.1l=h0d85af4_0
  - packaging=21.3=pyhd8ed1ab_0
  - pandas=1.3.5=py39h4d6be9b_0
  - pandoc=2.17.1.1=h694c41f_0
  - pandocfilters=1.5.0=pyhd8ed1ab_0
  - parso=0.8.3=pyhd8ed1ab_0
  - pathspec=0.9.0=pyhd8ed1ab_0
  - pcre=8.45=he49afe7_0
  - pexpect=4.8.0=pyh9f0ad1d_2
  - pickleshare=0.7.5=py_1003
  - pip=21.3.1=pyhd8ed1ab_0
  - platformdirs=2.5.1=pyhd8ed1ab_0
  - pluggy=1.0.0=py39h6e9494a_2
  - prometheus_client=0.13.1=pyhd8ed1ab_0
  - prompt-toolkit=3.0.27=pyha770c72_0
  - prompt_toolkit=3.0.27=hd8ed1ab_0
  - psutil=5.9.0=py39h89e85a6_0
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pure_eval=0.2.2=pyhd8ed1ab_0
  - py=1.11.0=pyh6c4a22f_0
  - pybind11-abi=4=hd8ed1ab_3
  - pycodestyle=2.8.0=pyhd8ed1ab_0
  - pycosat=0.6.3=py39h89e85a6_1009
  - pycparser=2.21=pyhd8ed1ab_0
  - pyflakes=2.4.0=pyhd8ed1ab_0
  - pygments=2.11.2=pyhd8ed1ab_0
  - pyopenssl=22.0.0=pyhd8ed1ab_0
  - pyparsing=3.0.7=pyhd8ed1ab_0
  - pyqt=5.12.3=py39h6e9494a_8
  - pyqt-impl=5.12.3=py39he44290a_8
  - pyqt5-sip=4.19.18=py39h15fb055_8
  - pyqtchart=5.12=py39he44290a_8
  - pyqtwebengine=5.12.1=py39he44290a_8
  - pyrsistent=0.18.1=py39h89e85a6_0
  - pysocks=1.7.1=py39h6e9494a_4
  - pytest=7.0.1=py39h6e9494a_0
  - python=3.9.7=h1248fe1_3_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python_abi=3.9=2_cp39
  - pytz=2021.3=pyhd8ed1ab_0
  - pyzmq=22.3.0=py39h7fec2f1_1
  - qt=5.12.9=h2a607e2_5
  - qtconsole=5.2.2=pyhd8ed1ab_1
  - qtconsole-base=5.2.2=pyhd8ed1ab_1
  - qtpy=2.0.1=pyhd8ed1ab_0
  - readline=8.1=h05e3726_0
  - reproc=14.2.3=h0d85af4_0
  - reproc-cpp=14.2.3=he49afe7_0
  - requests=2.27.1=pyhd8ed1ab_0
  - ruamel_yaml=0.15.80=py39h89e85a6_1006
  - scikit-learn=1.0.2=py39hd4eea88_0
  - scipy=1.7.3=py39h056f1c0_0
  - send2trash=1.8.0=pyhd8ed1ab_0
  - setuptools=59.8.0=py39h6e9494a_0
  - six=1.16.0=pyh6c4a22f_0
  - soupsieve=2.3.1=pyhd8ed1ab_0
  - sqlite=3.37.0=h23a322b_0
  - stack_data=0.2.0=pyhd8ed1ab_0
  - terminado=0.13.3=py39h6e9494a_0
  - testpath=0.6.0=pyhd8ed1ab_0
  - threadpoolctl=3.0.0=pyh8a188c0_0
  - tk=8.6.11=h5dbffcc_1
  - tokenize-rt=4.2.1=pyhd8ed1ab_0
  - tomli=1.2.2=pyhd8ed1ab_0
  - tornado=6.1=py39h89e85a6_2
  - tqdm=4.62.3=pyhd8ed1ab_0
  - traitlets=5.1.1=pyhd8ed1ab_0
  - typed-ast=1.5.2=py39h89e85a6_0
  - typing_extensions=4.0.1=pyha770c72_0
  - tzdata=2021e=he74cb21_0
  - urllib3=1.26.8=pyhd8ed1ab_1
  - wcwidth=0.2.5=pyh9f0ad1d_2
  - webencodings=0.5.1=py_1
  - wheel=0.37.1=pyhd8ed1ab_0
  - widgetsnbextension=3.5.2=py39h6e9494a_1
  - xz=5.2.5=haf1e3a3_1
  - yaml=0.2.5=h0d85af4_2
  - yaml-cpp=0.6.3=hb1e8313_4
  - zeromq=4.3.4=he49afe7_1
  - zipp=3.7.0=pyhd8ed1ab_1
  - zlib=1.2.11=h9173be1_1013
  - zstd=1.5.2=h582d3a0_0
  - pip:
    - bokeh==2.4.2
    - cloudpickle==2.0.0
    - colorama==0.4.3
    - dask==2022.3.0
    - deprecated==1.2.13
    - docutils==0.15.2
    - filelock==3.6.0
    - fsspec==2022.2.0
    - grpcio==1.44.0
    - jmespath==0.10.0
    - lightgbm==3.3.2.99
    - locket==0.2.1
    - msgpack==1.0.3
    - partd==1.2.0
    - pillow==9.0.1
    - protobuf==3.19.4
    - pyasn1==0.4.8
    - pyyaml==5.4.1
    - ray==1.10.0
    - redis==4.1.4
    - rsa==4.7.2
    - sf-hamilton==1.5.0
    - toolz==0.11.2
    - typing-inspect==0.7.1
    - wrapt==1.13.3
prefix: /Users/jlamb/mambaforge
```

</details>

Reply:
So given that, maybe this PR isn't introducing a regression. But I would still like to see both mine and @StrikerRUS 's recommended changes for the tests implemented before this PR is merged (as also requested in https://github.com/microsoft/LightGBM/pull/4827#issuecomment-1075779334).

In `test_sample_strategy_with_boosting`:

* please add comparisons of results across different configurations to each other, like 

```python
assert ret1 == ret2
assert ret1 != ret3
assert ret1 == ret5
```

* please add stricter tests on the value of evaluation metrics like `mean_squared_error()`. For example, instead of `ret < 4000`, something like the following (putting the expected value into a variable and re-using it, instead of repeating the same literal multiple times

```python
assert ret1 == pytest.approx(3954)
```

Those changes would give us much stronger confidence that future changes to LightGBM don't cause a regression in LightGBM's behavior under these different settings.

Reply:
@jameslamb Thanks a lot for your investigation! Agree with your proposals.

However, I still cannot reproduce the issue with the `master` branch on my machine.

Reply:
We would have to use `np.testing.assert_allclose` if we want to pass the test for GPU version. Since GPU version is non-deterministic and can have minor differences in the results.
![image](https://user-images.githubusercontent.com/14541765/172844250-52ae0144-8f5f-4bbf-8961-1c429690156d.png)

Reply:
I believe results for CPU and GPU should be identical with [`gpu_use_dp` param](https://lightgbm.readthedocs.io/en/latest/Parameters.html#gpu_use_dp).

Reply:
We've just got one independent report about the same issue: #5441.